### PR TITLE
hide dashcard when empty setting

### PR DESCRIPTION
**Pull request by alxnddr**
Related to https://github.com/metabase/metabase/issues/29175

### Description

Adds a new dashcard setting that defines if the card should be visible on a dashboard when results are empty.

### How to verify

1. Create a new dashboard with a few cards
2. Open the dashcards settings and toggle they should be hidden when no results
3. Connect filters to the dashcards
4. Enter filter values that lead to empty results
5. Ensure cards get hidden

### Demo

https://metaboat.slack.com/archives/C04D0E4837E/p1683940774587059

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30630)

### CODE REVIEW

File: frontend/src/metabase/dashboard/utils.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+export async function fetchDataOrError(dataPromise: Promise<unknown>) {
+  try {
+    return await dataPromise;
+  } catch (error) {
+    return { error };
+  }
+}
+
+export function getDatasetQueryParams(datasetQuery: any = {}) {
```
Comment (npfitz) [inline frontend/src/metabase/dashboard/utils.ts:154]: `any`

**Review by npfitz — CHANGES_REQUESTED** [review]
The only thing I noticed is that we don't seem to hide native queries when there are no results. Is this intentional? And there are a few instances of `any` that are just being called out incase they were forgotten/missed

image

**Comment by alxnddr** [comment]
@npfitz I could not reproduce it for native questions, can you please share more details on which param did you use? Regarding `any`s I renamed a js file to ts which contained many utils I did not change so because of that I did not refactor the entire file to add types. Anyway, I'll figure out and add types there

**Comment by npfitz** [comment]
Sure thing! So, I made a native question that looks like `Select * from Orders where CREATED_AT > {{CREATEDAT}}`, where `CREATEDAT` is a text input. Then you can try adding the parameter with a format like `2020-01-01`. I added that to the dashboard, connected the filter to it, and ran it. 

I also tried a native query like `Select * from PEOPLE where REGEXP_LIKE(NAME, {{name}}, 'i')` to search for people with specific names, and it also didn't disappear when no results were returned
image

**Comment by alxnddr** [comment]
For some reason, still works for me:
<video src="https://github.com/metabase/metabase/assets/14301985/a6571341-120a-4efd-b89f-0a9e685bdb1b" />

**Review by npfitz — APPROVED** [review]
Turns out my issue earlier was with the "chair to keyboard interface"... I forgot to turn the setting on for the dashcard. Looking good!

**Review by luizarakaki — APPROVED** [review]

**Pull request merged**